add menu

xavierua opened the pull request.

xavierua committed:
add menu

JustMeGaaRa commented on ToDo/MainMenu.csproj:
```diff
@@ -1,9 +1,9 @@
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net5.0</TargetFramework>
+    <RootNamespace>ToDo_Test_Project</RootNamespace>
```
Remove this line and let the runtime decide what the namespace should be

JustMeGaaRa commented on ToDo/Menu.cs:
```diff
@@ -1,112 +1,128 @@
-
-    }
-}
+﻿
```
Remove blank line

JustMeGaaRa commented on ToDo/Menu.cs:
```diff
@@ -1,112 +1,128 @@
+
+        public Menu(string title)
+        {
+            Title = title;
```
Check if string is null or empty and throw argument exception if so

JustMeGaaRa commented on ToDo/Menu.cs:
```diff
@@ -1,112 +1,128 @@
+
+        public void ExecuteItem(string key)
+        {
+            _items[key].Action();
```
What if there is no such item with the specified key? Add check

JustMeGaaRa commented on ToDo/Menu.cs:
```diff
@@ -1,112 +1,128 @@
+            _items[key].Action();
+        }
+
+        public string GetItems()
```
Why is this method returning string? You should use types, not string values here

JustMeGaaRa commented on ToDo/Menu.cs:
```diff
@@ -1,112 +1,128 @@
+        }
+
+
+        //private List<MenuItem> _itemsMenu = new();
```
Remove commented code, you have git history for this

JustMeGaaRa commented on ToDo/MenuItem.cs:
```diff
@@ -1,21 +1,21 @@
+    {
+        public MenuItem(string title, string button, Action action)
+        {
+            Title = title;
```
Check parameters if they are null and throw argument null exception if so

JustMeGaaRa commented on ToDo/MenuItem.cs:
```diff
@@ -1,21 +1,21 @@
+        public string Button { get; set; }
+
+        public Action Action { get; set; }
+
```
Remove blank line

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,191 @@
+
+namespace ToDo
+{
+    class Program
```
Add access modifier here. Use "internal" here

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,191 @@
+        private static int _numberOfTask;
+        private static Menu _menu = new("Main Menu");
+
+        static void Main(string[] args)
```
Use "public" access modifier here

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,191 @@
+
+            while (true)
+            {
+                Console.WriteLine(NavigationMenu());
```
Move "Console.WriteLine" to "NavigationMenu". Rename NamigationMenu to smth like "PrintNavigationMenu". Pass the _menu as parameter to PrintNavigationMenu

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,191 @@
+                //Console.WriteLine(_numberOfTasks);
+                Console.WriteLine(_positionInTask);
+
+                ShowTaskList();
```
Rename to "PrintTaskList". Pass task list as parameter to the method

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,191 @@
+{
+    class Program
+    {
+        private static TaskList _listOfTasks = new();
```
Rename _listOfTasks to _taskList. Move all static fields to Main method variables.

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,191 @@
+    {
+        private static TaskList _listOfTasks = new();
+        private static int _positionInTask = 0;
+        private static int _numberOfTask;
```
Rename _numberOfTask to _taskNumber

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,191 @@
+        {
+            if (status == "done")
+            {
+                _listOfTasks.GetTaskList()[_positionInTask].MarkAsDone();
```
A couple of issues here:
- No check if index is valid
- Getting item by index should be encapsulated into the TaskList itself
- GetTaskList is not a valid name, because you are calling a method on TaskList type

Introduce a method called "GetTaskByIndex" and include index check inside

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,191 @@
+
+        public static string NavigationMenu()
+        {
+            var menuInfo = _menu.GetItems();
```
You should've used StringBuilder here based on all task items

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,191 @@
+        {
+            List<Task> taskList = _listOfTasks.GetTaskList();
+
+            for (int i = 0; i < taskList.Count(); i++)
```
Use "Count" property instead of "Count()" method call

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,191 @@
+
+        public static int GetNumberOfTask()
+        {
+            return _listOfTasks.GetTaskList().Count;
```
Introduce a property "Count" on "taskList" type that returns the count of tasks from a List of tasks

JustMeGaaRa commented on ToDo/Task.cs:
```diff
@@ -1,48 +1,48 @@
+{
+    public class Task
+    {
+
```
Remove blank line here

JustMeGaaRa commented on ToDo/Task.cs:
```diff
@@ -1,48 +1,48 @@
+    {
+
+        private static int ID = 0;
+        public Task(string title)
```
Add blank line here

JustMeGaaRa commented on ToDo/Task.cs:
```diff
@@ -1,48 +1,48 @@
+            CreatedAt = DateTime.UtcNow;
+            ID += 1;
+        }
+        public int TaskID { get; private set; }
```
Add blank line here

JustMeGaaRa commented on ToDo/Task.cs:
```diff
@@ -1,48 +1,48 @@
+            CompletedAt = null;
+        }
+
+        public void AddDescription(string description)
```
Rename to "SetDechsription" or "ChangeDescription" you do not add the description if there is a value already

xavierua committed:
remove redanted files

xavierua committed:
Remove commented code, add check parametres, rename variables and methods

JustMeGaaRa commented on ToDo/Menu.cs:
```diff
@@ -1,112 +1,49 @@
+            {
+                _items[key].Action();
+            }
+            
```
Remove blank line

JustMeGaaRa commented on ToDo/Menu.cs:
```diff
@@ -1,112 +1,49 @@
+            }
+            return itemsList;
+        }
+        public bool CheckIsKeyValid(string key)
```
This method should not be public, make it private

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,190 @@
+        private static int _position = 0;
+        private static TaskList _taskList = new();
+        private static Menu _menu = new("Main Menu");
+        public static void Main(string[] args)
```
Add a blank line

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,190 @@
+                Console.Clear();
+            }
+        }
+        public static void ExecuteAddNewTask()
```
Add a blank line

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,190 @@
+            Task newTask = new(taskTitle);
+            _taskList.AddToList(newTask);
+        }
+        public static void ExecuteChangeDescription()
```
Add a blank line

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,190 @@
+            string userInput = Console.ReadLine();
+            task.ChangeDescription(userInput);
+        }
+        public static void ExecuteStatusTask(int index, string status)
```
Add a blank line

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,190 @@
+                Console.WriteLine("The task list is empty!");
+            }
+        }
+        public static void ExecuteDeleteTask()
```
Add a blank line

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,190 @@
+                Console.WriteLine("The task list is empty!");
+            }
+        }
+        public static void ExecuteChangePosition(int index, string direction)
```
Add a blank line

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,190 @@
+                    _position += 1;
+                }
+            }
+
```
Remove blank line

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,190 @@
+            }
+
+        }
+        public static void PrintMenu(Menu menu)
```
Add a blank line

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,190 @@
+                .Append('=', 40);
+            Console.WriteLine(menuInfo.ToString());
+        }
+        public static void PrintTaskList(TaskList taskList)
```
Add a blank line

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,190 @@
+                }
+            }
+        }
+        public static bool IsValid(string userInput)
```
Add a blank line

JustMeGaaRa commented on ToDo/Program.cs:
```diff
@@ -1,200 +1,190 @@
+
+            return userInput != string.Empty || userInput != null;
+        }
+        public static bool IsPositionValid(int position, string wayToMove)
```
Add a blank line

JustMeGaaRa commented on ToDo/Task.cs:
```diff
@@ -1,48 +1,41 @@
+        {
+            return $"Title: {Title}\n\tDescription: {Description}\n\tIs Done: {IsDone}\n\tCreated At: {CreatedAt}\n\tCompleted At: {CompletedAt}\n";
+        }
+        public void MarkAsDone()
```
Add a blank line

JustMeGaaRa commented on ToDo/Task.cs:
```diff
@@ -1,48 +1,41 @@
+            IsDone = true;
+            CompletedAt = DateTime.UtcNow;
+        }
+        public void MarkAsUndone()
```
Add a blank line

JustMeGaaRa commented on ToDo/Task.cs:
```diff
@@ -1,48 +1,41 @@
+            IsDone = false;
+            CompletedAt = null;
+        }
+        public void ChangeDescription(string description)
```
Add a blank line

JustMeGaaRa commented on ToDo/TaskList.cs:
```diff
@@ -1,34 +1,48 @@
+        {
+            _tasks.Add(task);
+        }
+        public bool DeleteTask(int taskId)
```
Add a blank line

JustMeGaaRa commented on ToDo/TaskList.cs:
```diff
@@ -1,34 +1,48 @@
+                ? false
+                : _tasks.Remove(_tasks[taskId]);
+        }
+        public List<Task> ListOfTask()
```
Replace with a property like "public IReadOnlyCollection<Task> Tasks { get { return _tasks; } }"

JustMeGaaRa commented on ToDo/TaskList.cs:
```diff
@@ -1,34 +1,48 @@
+        {
+            return _tasks;
+        }
+        public Task GetTaskByIndex(int index)
```
Add a blank line

JustMeGaaRa commented on ToDo/TaskList.cs:
```diff
@@ -1,34 +1,48 @@
+        {
+            return _tasks[index];
+        }
+        public bool IsIndexValid(int index)
```
Add a blank line

JustMeGaaRa commented on ToDo/TaskList.cs:
```diff
@@ -1,34 +1,48 @@
+            }
+            return false;
+        }
+        public bool IsEmpty()
```
Add a blank line

xavierua committed:
add blank, move "taskList" and "menu" to "Main" function

- add new parameters to methods in Program class;
-  change access modifier to private in "CheckIsKeyValid" method;
